test/integration/authorize/list_app_keys_test.rb

```diff
@@ -0,0 +1,132 @@
+require File.expand_path(File.dirname(__FILE__) + '/../../test_helper')
+
+class AuthorizeListAppKeysExtensionTest < Test::Unit::TestCase
+  include TestHelpers::Fixtures
+  include TestHelpers::Integration
+  include TestHelpers::Extensions
+
+  def setup
+    @storage = Storage.instance(true)
+    @storage.flushdb
+
+    Resque.reset!
+    Memoizer.reset!
+
+    setup_provider_fixtures
+
+    @application = Application.save(service_id: @service.id,
+                                    id: next_id,
+                                    state: :active,
+                                    plan_id: @plan_id,
+                                    plan_name: @plan_name)
+
+    @keys = 10.times.map { |i| "key_#{i}" }
+
+    @keys.each do |k|
+      @application.create_key k
+    end
+
+    @application_nokeys = Application.save(service_id: @service.id,
+                                           id: next_id,
+                                           state: :active,
+                                           plan_id: @plan_id,
+                                           plan_name: @plan_name)
+  end
+
+  test 'calling authorize without the list_app_keys extension does not output the application keys section' do
+    get '/transactions/authorize.xml',
+        { provider_key: @provider_key, app_id: @application.id }
+
+    xml_resp = Nokogiri::XML(last_response.body)
+    assert xml_resp.at('app_keys').nil?
+  end
+
+  test 'calling authorize with the list_app_keys extension outputs the application keys section even if no app keys exist' do
+    get '/transactions/authorize.xml',
+        { provider_key: @provider_key, app_id: @application_nokeys.id },
+        'HTTP_3SCALE_OPTIONS' => Extensions::LIST_APP_KEYS
+
+    xml_resp = Nokogiri::XML(last_response.body)
+    app_keys = xml_resp.at 'app_keys'
+    assert !app_keys.nil?
+    assert_equal @application_nokeys.id, app_keys['app']
+    assert_equal @service.id, app_keys['svc']
+
+    keys = xml_resp.search("app_keys key").map { |n| n[:id] }
+
+    assert keys.empty?
+  end
+
+  test 'calling authorize with the list_app_keys extension lists the application keys even if no app_key is specified' do
+    get '/transactions/authorize.xml',
+        { provider_key: @provider_key, app_id: @application.id },
+        'HTTP_3SCALE_OPTIONS' => Extensions::LIST_APP_KEYS
+
+    xml_resp = Nokogiri::XML(last_response.body)
+    app_keys = xml_resp.at 'app_keys'
+    assert !app_keys.nil?
+    assert_equal @application.id, app_keys['app']
+    assert_equal @service.id, app_keys['svc']
+
+    keys = xml_resp.search("app_keys key").map { |n| n[:id] }
+
+    assert_equal @keys, keys.sort
```

Comment:
sort `@keys` too, so we don't have to look at the beginning of the file to check whether it's sorted or not.

Reply:
Hm, this one is quite subjective: this can only consistently succeed if `@keys` is sorted, that's kind of the premise, and sorting something that can only be sorted to succeed in the general case means more repetitions and more runtime for tests :).

That said... if you still think it's worth it to repeat it... just insist on it here (technically: _repeat_ it to me ;). :neckbeard:

Reply:
You decided that `@keys` is sorted, but for your tests you could have chosen any other thing (adapting the assert).

My point is, when comparing arrays I think it's better for readability to sort both because that works always. If it's very clear that they're sorted this might not be needed.

When you get to this line of the assert and you see that `@keys` is not sorted, now you have to go back to the line that defines it to see if it is or not :neckbeard: